src/pudl/extract/eiaaeo.py

```diff
@@ -0,0 +1,252 @@
+"""Extract EIA AEO data from the bulk JSON."""
+
+import io
+import itertools
+import re
+from collections.abc import Iterable
+
+import networkx as nx
+import pandas as pd
+from dagster import (
+    AssetCheckResult,
+    AssetExecutionContext,
+    AssetOut,
+    Output,
+    asset_check,
+    multi_asset,
+)
+from pandera import DataFrameModel, Field
+from pandera.dtypes import Timestamp
+from pydantic import BaseModel
+
+import pudl.logging_helpers
+
+logger = pudl.logging_helpers.get_logger(__name__)
+
+
+class Category(BaseModel):
+    """Describe the structure of how series are related."""
+
+    category_id: int
+    parent_category_id: int
+    name: str
+    notes: str
+    childseries: list[str]
+
+
+class Series(BaseModel):
+    """Contain the actual data, and each timeseries' metadata."""
+
+    series_id: str
+    name: str
+    last_updated: str
+    units: str | None = None
+    data: list[tuple[str, str | float]]  # time, value
+
+
+class AEOTable(DataFrameModel):
+    """Data schema for a raw AEO table."""
+
+    date: Timestamp = Field(coerce=True)
+    value: str = Field(coerce=True)
+    units: str = Field(coerce=True)
+    series_name: str = Field(coerce=True)
+    category_name: str = Field(coerce=True)
+    case: str = Field(coerce=True)
+
+
+class AEOTaxonomy:
+    """Container for *all* the information in one AEO report."""
+
+    def __init__(self, records: Iterable[str]):
+        """Load AEO JSON records into a graph datastructure.
+
+        Args:
+            records: the strings that contain an AEO report to parse.
+
+        """
+        categories, series = self.__load_records(records)
+        self.graph = self.__generate_graph(categories, series)
+        root = 4949903
+        self.__cases = set(self.graph.successors(root))
```

Comment:
So there's a structural requirement that the root node never has any child series, and that the top level categories always represent the different cases / scenarios?